### PR TITLE
virttest.qemu_monitor.py: Update params for live_snapshot().

### PR DESCRIPTION
**Pull request by CongLi**
Signed-off-by: Cong Li <coli@redhat.com>

id: 1321820

### CODE REVIEW

File: virttest/qemu_monitor.py [hunk 1]
```diff
@@ -1835,19 +1835,22 @@ def migrate_set_downtime(self, value):
         args = {"value": val}
         return self.cmd("migrate_set_downtime", args)
 
-    def live_snapshot(self, device, snapshot_file, snapshot_format="qcow2"):
+    def live_snapshot(self, device, snapshot_file, snapshot_format="qcow2",
+                      snapshot_mode="absolute-paths"):
```
Comment (xutian) [inline virttest/qemu_monitor.py]: It looks this valu should be 'absolute-path'

Comment (xutian) [inline virttest/qemu_monitor.py]: Oh, you are right. qmp-commands.hx:          (NewImageMode, optional, default 'absolute-paths')

Comment (xutian) [inline virttest/qemu_monitor.py]: live_snapshot function in qemu_vm module should be update too
def live_snapshot(self, base_file, snapshot_file,
                      snapshot_format="qcow2"):

**Comment by CongSmile** [comment]
@xutian 

The commit info has updated, could you help review it again?

Thanks.

**Comment by xutian** [comment]
Hi you should update HMP interface too.

**Comment by CongSmile** [comment]
@xutian 

I have added notes in the commit log.
Note: HMP does not support option "mode".
snapshot_blkdev [-n] device [new-image-file] [format]

Could you double check it?

Thanks.

**Comment by CongSmile** [comment]
Result:

1. HMP: 
13:16:19 DEBUG| Send command: snapshot_blkdev drive_image1 /usr/share/avocado/data/avocado-vt/images/RHEL-Server-7.3-64-virtio-snap.qcow2 qcow2
13:16:19 INFO | [qemu output] Formatting '/usr/share/avocado/data/avocado-vt/images/RHEL-Server-7.3-64-virtio-snap.qcow2', fmt=qcow2 size=21474836480 backing_file=/usr/share/avocado/data/avocado-vt/images/RHEL-Server-7.3-64-virtio.qcow2 backing_fmt=qcow2 encryption=off cluster_size=65536 lazy_refcounts=off refcount_bits=16
13:16:24 INFO | PASS smp_8.2048m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.snapshot_test

2. QMP:
13:04:58 DEBUG| Send command: {'execute': 'blockdev-snapshot-sync', 'arguments': {'device': u'drive_image1', 'snapshot-file': '/usr/share/avocado/data/avocado-vt/images/RHEL-Server-7.3-64-virtio-snap.qcow2', 'mode': 'absolute-paths', 'format': 'qcow2'}, 'id': 'q0jrcaVt'}
13:04:58 INFO | [qemu output] Formatting '/usr/share/avocado/data/avocado-vt/images/RHEL-Server-7.3-64-virtio-snap.qcow2', fmt=qcow2 size=21474836480 backing_file=/usr/share/avocado/data/avocado-vt/images/RHEL-Server-7.3-64-virtio.qcow2 backing_fmt=qcow2 encryption=off cluster_size=65536 lazy_refcounts=off refcount_bits=16
13:05:03 INFO | PASS smp_8.2048m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.snapshot_test

File: virttest/qemu_monitor.py [hunk 2]
```diff
@@ -849,18 +849,23 @@ def set_link(self, name, up):
...
+        :param device: the name of the device to generate the snapshot from.
+        :param snapshot_file: the target of the new image. A new file will be created.
+        :param kwargs: optional keyword arguments to pass to func
 
         :return: The response to the command
         """
-        cmd = ("snapshot_blkdev %s %s %s" %
-               (device, snapshot_file, snapshot_format))
+        device = device
+        snapshot_file = snapshot_file
```
Comment (suqinhuang) [inline virttest/qemu_monitor.py]: why do we need thest two lines

Comment (PyLearner) [inline virttest/qemu_monitor.py]: hi, coli ,why will you  define the variable inside the function again ? thanks.

Comment (CongSmile) [inline virttest/qemu_monitor.py]: According to the definition in the qemu code, params "device" and "snapshot-file" are must.
Others are optional.
kwargs is define those optional keywork arguments.

:param kwargs: optional keyword arguments to pass to func

# @BlockdevSnapshot
# @device:  the name of the device to generate the snapshot from.
# @snapshot-file: the target of the new image. A new file will be created.
# @format: #optional the format of the snapshot image, default is 'qcow2'.
# @mode: #optional whether and how QEMU should create a new image, default is
#        'absolute-paths'.

File: virttest/qemu_monitor.py [hunk 3]
```diff
@@ -1835,20 +1840,27 @@ def migrate_set_downtime(self, value):
...
-    def live_snapshot(self, device, snapshot_file, snapshot_format="qcow2"):
+    def live_snapshot(self, device, snapshot_file, **kwargs):
         """
         Take a live disk snapshot.
 
-        :param device: device id of base image
-        :param snapshot_file: image file name of snapshot
-        :param snapshot_format: image format of snapshot
+        :param device: the name of the device to generate the snapshot from.
+        :param snapshot_file: the target of the new image. A new file will be created.
```
Comment (vivianQizhu) [inline virttest/qemu_monitor.py]: A new file will be created if 'mode' is 'absolute_paths'

**Comment by vivianQizhu** [comment]
Test with 'vm.monitor.live_snapshot(device, snapshot_name, format="qcow2", node_name="node1", snapshot_node_name="node2")'
Checking with the argument below, I am not sure if it missed the default value 'mode'? @CongLi, is this what you intends in your code? Thanks

10:03:41 DEBUG| Send command: {'execute': 'blockdev-snapshot-sync', 'arguments': {'snapshot-node-name': 'node2', 'node-name': 'node1', 'format': 'qcow2', 'device': u'drive_image1', 'snapshot-file': 'live_snapshot_img'}, 'id': 'h52rBZni'}
10:04:56 INFO | PASS 0-smp_4.4096m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.RHEL.7.3.x86_64.io-github-autotest-qemu.live_snapshot.with_runtime

**Comment by CongLi** [comment]
@vivianQizhu 

According to the qemu doc, mode is an optional param.
@BlockdevSnapshot
@device: the name of the device to generate the snapshot from.
@snapshot-file: the target of the new image. A new file will be created.
@format: #optional the format of the snapshot image, default is 'qcow2'.
@mode: #optional whether and how QEMU should create a new image, default is 'absolute-paths'.

**Comment by vivianQizhu** [comment]
LGTM.  Acked-by: Qianqian Zhu qizhu@redhat.com

**Comment by CongLi** [comment]
@xutian 

Could you help review this one?

Thanks.

File: virttest/qemu_monitor.py (hunk 3 above)
Comment (CongLi) [inline virttest/qemu_monitor.py]: HMP does not support option 'mode'

**Comment by CongLi** [comment]
@xutian

Could you help review this one?

Thanks.

File: virttest/qemu_monitor.py [hunk 4]
```diff
@@ -1834,20 +1845,34 @@ def migrate_set_downtime(self, value):
...
+        Optional keyword:
+        @format: #optional the format of the snapshot image, default is 'qcow2'.
         """
         args = {"device": device,
-                "snapshot-file": snapshot_file,
-                "format": snapshot_format}
-        return self.cmd("blockdev-snapshot-sync", args)
+                "snapshot-file": snapshot_file}
+        kwargs.update(args)
+        if not kwargs['format']:
```
Comment (xutian) [inline virttest/qemu_monitor.py]: if format not in kwargs:
     xxx
mode same as format.

Comment (CongLi) [inline virttest/qemu_monitor.py]: Updated.
Thanks @xutian

File: virttest/qemu_monitor.py [hunk 5]
```diff
@@ -1834,20 +1845,34 @@ def migrate_set_downtime(self, value):
...
+            kwargs.update({"format": "qcow2"})
+        if not kwargs['mode']:
+            kwargs.update({"mode": "absolute-paths"})
+        for key, value in kwargs.items():
+            new_key = key.replace('_', '-')
+            del kwargs[key]
+            args = {new_key: value}
+            kwargs.update(args)
```
Comment (PandaWei) [inline virttest/qemu_monitor.py]: Is there any key with '_' ?
If not, why we need to replace '_'?

Comment (CongLi) [inline virttest/qemu_monitor.py]: Thanks @PandaWei 

Removed.

**Comment by xutian** [comment]
@CongLi  can you show me your test results with last update.

thanks,
Xu

**Comment by CongLi** [comment]
10:27:10 INFO | PASS 1-smp_8.2048m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.Win2012.x86_64.r2.io-github-autotest-qemu.live_snapshot_basic
10:27:09 DEBUG| Send command: {'execute': 'blockdev-snapshot-sync', 'arguments': {'device': u'drive_image1', 'snapshot-file': '/usr/share/avocado/data/avocado-vt/images/live_snapshot_img', 'mode': 'absolute-paths', 'format': 'qcow2'}, 'id': 'KmHzV4LX'}


10:29:05 INFO | PASS 2-smp_8.2048m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.Win2012.x86_64.r2.io-github-autotest-qemu.live_snapshot_basic.existing_mode
10:27:14 DEBUG| Send command: {'execute': 'blockdev-snapshot-sync', 'arguments': {'device': u'drive_image1', 'snapshot-file': '/usr/share/avocado/data/avocado-vt/images/win2012-64r2-virtio-snap.qcow2', 'mode': 'existing', 'format': 'qcow2'}, 'id': '6U0r7fd4'}

10:29:20 INFO | PASS 3-smp_8.2048m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.Win2012.x86_64.r2.io-github-autotest-qemu.live_snapshot_basic.node_name
10:29:08 DEBUG| Send command: {'execute': 'blockdev-snapshot-sync', 'arguments': {'node-name': 'node1', 'mode': 'absolute-paths', 'format': 'qcow2', 'device': u'drive_image1', 'snapshot-node-name': 'node2', 'snapshot-file': '/usr/share/avocado/data/avocado-vt/images/live_snapshot_img'}, 'id': 'WY2cHAMi'}

**Comment by xutian** [comment]
@CongLi   Ci check failed, please fix it

**Comment by CongLi** [comment]
@xutian Updated, thanks.

File: virttest/qemu_monitor.py [hunk 6]
```diff
@@ -849,18 +849,25 @@ def set_link(self, name, up):
...
+            format: the format of the snapshot image, default is 'qcow2'.
 
-        :return: The response to the command
+        :return: The response to the command.
         """
-        cmd = ("snapshot_blkdev %s %s %s" %
-               (device, snapshot_file, snapshot_format))
+        device = device
+        snapshot_file = snapshot_file
+        if 'format' in kwargs:
```
Comment (xutian) [inline virttest/qemu_monitor.py]: cmd = "snapshot_blkdev  %s %s" % (device, snapshot_file)
if 'format' in kwargs:
     cmds += " %s" % kwargs['format']
return self.cmd(cmd)

File: virttest/qemu_monitor.py [hunk 7]
```diff
@@ -849,18 +849,23 @@ def set_link(self, name, up):
...
-        :return: The response to the command
+        :return: The response to the command.
         """
-        cmd = ("snapshot_blkdev %s %s %s" %
-               (device, snapshot_file, snapshot_format))
+        device = device
+        snapshot_file = snapshot_file
+        cmd = "snapshot_blkdev %s %s" % (device, snapshot_file)
+        if 'format' in kwargs:
+            cmd += "%s" % kwargs['format']
```
Comment (xutian) [inline virttest/qemu_monitor.py]: leak a space.

Comment (CongLi) [inline virttest/qemu_monitor.py]: HMP:
18:12:40 DEBUG| (monitor hmp1) Sending command 'snapshot_blkdev drive_image1 live_snapshot_img qcow2'
18:12:40 DEBUG| Send command: snapshot_blkdev drive_image1 live_snapshot_img qcow2

18:13:25 INFO | PASS 1-smp_8.2048m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.RHEL.7.2.x86_64.io-github-autotest-qemu.live_snapshot.with_reboot

**Comment by CongLi** [comment]
@xutian Could you help review it? Thx.

**Comment by xutian** [comment]
looks ok to me, thanks @CongLi and @vivianQizhu . merged

**Pull request merged**